crates/storage/db/src/implementation/mdbx/cursor.rs

```diff
@@ -24,13 +25,44 @@ pub type CursorRW<'tx, T> = Cursor<'tx, RW, T>;
 #[derive(Debug)]
 pub struct Cursor<'tx, K: TransactionKind, T: Table> {
     /// Inner `libmdbx` cursor.
-    pub inner: reth_libmdbx::Cursor<'tx, K>,
-    /// Table name as is inside the database.
-    pub table: &'static str,
-    /// Phantom data to enforce encoding/decoding.
-    pub _dbi: std::marker::PhantomData<T>,
+    pub(crate) inner: reth_libmdbx::Cursor<'tx, K>,
     /// Cache buffer that receives compressed values.
-    pub buf: Vec<u8>,
+    buf: Vec<u8>,
+    /// Whether to record metrics or not.
+    with_metrics: bool,
+    /// Phantom data to enforce encoding/decoding.
+    _dbi: PhantomData<T>,
+}
+
+impl<'tx, K: TransactionKind, T: Table> Cursor<'tx, K, T> {
+    pub(crate) fn new_with_metrics(
+        inner: reth_libmdbx::Cursor<'tx, K>,
+        with_metrics: bool,
+    ) -> Self {
+        Self { inner, buf: Vec::new(), with_metrics, _dbi: PhantomData }
+    }
+
+    /// If `self.with_metrics == true`, measure the time it takes to execute the closure and record
+    /// a metric with the provided operation.
+    ///
+    /// Otherwise, just execute the closure.
+    fn execute_with_operation_metric<R>(
+        &mut self,
+        operation: Operation,
+        f: impl FnOnce(&mut Self) -> R,
+    ) -> R {
+        if self.with_metrics {
+            let start = Instant::now();
+            let result = f(self);
+            let duration = start.elapsed();
```

Comment:
do we know how often this is called?

skeptical that we need to this for all operations like upsert/put etc.., this could be excessive doing a clock syscall per operation?

Reply:
good point, it's called very often and may cause a significant perf hit. I changed it to measure time only on operation with large values of `put/upsert/insert/append/append_dup` operations because these are the ones we're interested in.